contracts/0.8.9/WithdrawlQueueBase.sol

```diff
@@ -0,0 +1,344 @@
+        uint256 indexInQueue;
+        /// @notice discount factor with 1e27 precision (0 - 100% discount, 1e27 - means no discount)
+        uint96 discountFactor;
+    }
+
+    /// @notice Emitted when a new withdrawal request enqueued
+    /// @dev Contains both stETH token amount and its corresponding shares amount
+    event WithdrawalRequested(
+        uint256 indexed requestId,
+        address indexed requestor,
+        address indexed recipient,
+        uint256 amountOfStETH,
+        uint256 amountOfShares
+    );
+    event WithdrawalClaimed(uint256 indexed requestId, address indexed receiver, address initiator);
+
+    error ZeroRecipientAddress();
+    error ZeroRequestId();
+    error SenderExpected(address _recipient, address _msgSender);
+    error InvalidRequestId(uint256 _requestId);
+    error NotEnoughEther();
+    error RequestNotFinalized(uint256 _requestId);
+    error RequestAlreadyClaimed();
+    error InvalidHint(uint256 _hint);
+    error CantSendValueRecipientMayHaveReverted();
+    error SafeCastValueDoesNotFit64Bits();
+    error SafeCastValueDoesNotFit96Bits();
+    error SafeCastValueDoesNotFit128Bits();
+
+    /// @notice queue for withdrawal requests
+    /// @dev mapping for better upgradability
+    mapping(uint256 => WithdrawalRequest) internal queue;
+
+    /// @notice length of the queue
+    uint256 public lastRequestId = 0;
+
+    /// @notice length of the finalized part of the queue. Always <= `requestCounter`
+    uint256 public lastFinalizedRequestId = lastRequestId;
+
+    /// @notice finalization discount history
+    /// @dev mapping for better upgradability
+    mapping(uint256 => Discount) internal discountHistory;
+
+    /// @notice size of discount history
+    uint256 public lastDiscountIndex = 0;
+
+    /// @notice amount of ETH on this contract balance that is locked for withdrawal and waiting for claim
+    uint128 public lockedEtherAmount = 0;
+
+    /// @notice withdrawal requests mapped to the recipients
+    mapping(address => uint256[]) internal requestsByRecipient;
+
+    function _initializeQueue() internal {
+        // setting dummy zero structs in discountHistory and queue beginning
+        // to avoid uint underflows and related if-branches
+        queue[lastRequestId] = WithdrawalRequest(0, 0, payable(0), _toUint64(block.number), true);
+        discountHistory[lastDiscountIndex] = Discount(lastRequestId, NO_DISCOUNT);
+    }
+
+    /// @notice return the number of unfinalized requests in the queue
+    function unfinalizedRequestNumber() external view returns (uint256) {
+        return lastRequestId - lastFinalizedRequestId;
+    }
+
+    /// @notice Returns the amount of stETH yet to be finalized
+    function unfinalizedStETH() external view returns (uint256) {
+        return queue[lastRequestId].cumulativeStETH - queue[lastFinalizedRequestId].cumulativeStETH;
+    }
+
+    /// @notice Returns all withdrawal requests placed for the `_recipient` address
```

Comment:
maybe not `placed`, but `belonging to` since the position is transferrable

Reply:
Should we change recipient to owner, btw? It fits better as an abstraction here, as I see.

Reply:
hmm, seems legit

Reply:
Will do it later because rn it will break other PRs